packages/core/src/components/fields/ArrayField.tsx

```diff
@@ -500,6 +500,26 @@ class ArrayField<T = any, S extends StrictRJSFSchema = RJSFSchema, F extends For
         const itemErrorSchema = errorSchema ? (errorSchema[index] as ErrorSchema<T[]>) : undefined;
         const itemIdPrefix = idSchema.$id + idSeparator + index;
         const itemIdSchema = schemaUtils.toIdSchema(itemSchema, itemIdPrefix, itemCast, idPrefix, idSeparator);
+
+        // Compute the item UI schema - either use the static object or call the function
+        let itemUiSchema: UiSchema<T[], S, F> | undefined;
+        if (typeof uiSchema.items === 'function') {
+          try {
+            // Call the function with item data, index, and form context
+            // TypeScript now correctly infers the types thanks to the ArrayElement type in UiSchema
+            const result = uiSchema.items(item, index, formContext);
+            // Only use the result if it's truthy
+            itemUiSchema = result as UiSchema<T[], S, F>;
+          } catch (e) {
+            console.error(`Error executing dynamic uiSchema.items function for item at index ${index}:`, e);
+            // Fall back to undefined to allow the field to still render
+            itemUiSchema = undefined;
+          }
+        } else {
+          // Static object case - preserve undefined to maintain backward compatibility
+          itemUiSchema = uiSchema.items as UiSchema<T[], S, F> | undefined;
+        }
```

Comment:
Can you DRY this and other identical implementation into a helper method on the class?

Reply:
Will do.